pylint/extensions/dict_init_mutate.py

```diff
@@ -0,0 +1,118 @@
+
+if TYPE_CHECKING:
+    from pylint.lint.pylinter import PyLinter
+
+
+def stringify_dict_items(node: nodes.NodeNG) -> str:
+    try:
+        key = node.targets[0].slice.value
+        value = node.value.value
+    except AttributeError:
+        # More complex cases will be handled later on
+        return ""
+
+    if isinstance(value, str):
+        value = f"'{value}'"
+    return f"'{key}': {value}"
+
+
+class DictInitMutateChecker(BaseChecker):
+    name = "dict-init-mutate"
+    msgs = {
+        "C3301": (
+            "Declare all known key/values when initializing the dictionary with '%s'",
+            "dict-init-mutate",
+            "Dictionaries can be initialized with a single statement "
+            "using dictionary literal syntax.",
+        )
+    }
+
+    @only_required_for_messages("dict-init-mutate")
+    def visit_assign(self, node: nodes.Assign) -> None:
+        """
+        Detect dictionary mutation immediately after initialization.
+
+        At this time, detecting nested mutation is not supported.
+        """
+        if not isinstance(node.value, nodes.Dict):
+            return
+
+        dict_name = node.targets[0]
+        if len(node.targets) != 1 or not isinstance(dict_name, nodes.AssignName):
+            return
+
+        first_sibling = node.next_sibling()
+        if (
+            not first_sibling
+            or not isinstance(first_sibling, nodes.Assign)
+            or len(first_sibling.targets) != 1
+        ):
+            return
+
+        sibling_target = first_sibling.targets[0]
+        if not isinstance(sibling_target, nodes.Subscript):
+            return
+
+        sibling_name = sibling_target.value
+        if not isinstance(sibling_name, nodes.Name):
+            return
+
+        if sibling_name.name == dict_name.name:
+            siblings = self._get_dict_mutating_siblings(first_sibling)
+            suggestion = self._make_dict_suggestion(siblings, dict_name.name)
+            self.add_message(
+                "dict-init-mutate", node=node, args=(suggestion,), confidence=HIGH
+            )
+
+    def _get_dict_mutating_siblings(
+        self, node: nodes.Assign, all_siblings: List[nodes.Assign] | None = None
+    ) -> List[nodes.Assign]:
+        """Get all sibling nodes that assign values to a dict."""
+        if all_siblings is None:
+            all_siblings = []
+
+        sibling = node.next_sibling()
+        if (
+            not sibling
+            or not isinstance(sibling, nodes.Assign)
+            or len(sibling.targets) != 1
+        ):
+            return [node] + all_siblings
+
+        sibling_target = sibling.targets[0]
+        if not isinstance(sibling_target, nodes.Subscript):
+            return [node] + all_siblings
+
+        sibling_name = sibling_target.value
+        if not isinstance(sibling_name, nodes.Name):
+            return [node] + all_siblings
+
+        return [node] + self._get_dict_mutating_siblings(sibling, all_siblings)
```

Comment:
Much as I love recursion, this will crash Pylint if the dictionary is big enough:

```python
config = {}
config[0] = None
config[1] = None
...  # 2 ... 981
config[982] = None
config[983] = None
```

Hopefully nobody would write something like that by hand! But it could be an issue with generated files.

Maybe it would be best to limit the suggested replacement to checking only a few items (say, three). Then if there are more than that, add an elipsis: `Declare all known key/values when initializing the dictionary with 'config = {'dir': 'bin', 'user': 'me', 'workers': 5, ...}'`

This could be done in a later PR.

Reply:
very good point, I actually I'm really not loving the additional suggestion. I would vote to simply `Declare all known key/values when initializing the dictionary` and not include the constructing dictionary help. It adds difficult to read and maintain code that doesn't add THAT much value. @Pierre-Sassoulas what do you think?

Reply:
That would be awesome. I'll update the PR to leave the general suggestion and open a new issue to add the improvement suggestion.